### ISSUE TITLE
Indicating un-resizable windows - not sure if this is actually possible!

### ISSUE THREAD

**Issue by spl237** [labwc/labwc#1294]
I think this is the one remaining difference between window header bars on openbox and on labwc.

If a window is set to not be resizable, for example in GTK, then in X/Openbox, the window header buttons reflect this - the "maximize" button is greyed-out and non-functional, and the window resize handles are not shown when the mouse pointer is on the window edges.

In labwc, this doesn't happen - all windows are always resizable. Even if a GTK window is set to not be resizable, it has resize handles (which work), and can be maximized.

In wayfire, this is partly addressed by the use of client-side decorations, whereby a GTK window gets a GTK headerbar, and that knows whether or not the window is resizable and sets buttons etc accordingly. But even then, it only works in client-side decorated windows - any window which is server-side decorated can be resized.

As most (all?) windows in labwc are server-side decorated, is there any way for the header bar to determine whether the client window is set to non-resizable? I assume it would require a wayland protocol to pass the information up from the client, but I don't even know if such a protocol exists - I have heard that it may be in the xdg-shell or gtk-shell protocol extensions, but I haven't looked at them yet.

I'm happy to have a go at implementing this myself in terms of the headerbar code, but I don't know how to get the required information from the client - does anyone have any knowledge or experience in this area?

**Comment by Consolatis** [comment]
You could set labwc to prefer CSD rather than SSD for the decorations: https://labwc.github.io/labwc-config.5.html#entry_core_decoration

As for clients preventing resizing.. that in itself is a discussion that is not that easy. My personal (e.g. as a user) preference would be that the user should always be able to resize windows and the window should not be dictating if it can be resized or not. This is especially important on small screens and applications preventing resize of dialog windows and similar.

As for the gtk-shell protocol: we will usually not implement private protocols (e.g. anything not in [wlr-protocols](https://gitlab.freedesktop.org/wlroots/wlr-protocols/) or [wayland-protocols](https://gitlab.freedesktop.org/wayland/wayland-protocols/)).

**Comment by spl237** [comment]
> You could set labwc to prefer CSD rather than SSD for the decorations: https://labwc.github.io/labwc-config.5.html#entry_core_decoration

SSD fits better with the way our hardware draws to the screen, so we'd prefer to use SSD from an efficiency point of view.

> As for clients preventing resizing.. that in itself is a discussion that is not that easy. My personal (e.g. as a user) preference would be that the user should always be able to resize windows and the window should not be dictating if it can be resized or not. This is especially important on small screens and applications preventing resize of dialog windows and similar.

I see your point, but I guess the counter-argument is that if an app developer has decided a window shouldn't be resized, then resizing will probably end up with some widgets looking wrong or the window layout becoming strange. I think that as long as the likes of GTK allow windows to be set as non-resizable, that ought to be respected, as the consequences of not respecting it can be unpredictable!

**Comment by spl237** [comment]
Looking at gtk-shell, it looks as if it offers quite a few features we would want - in addition to supporting resizing restrictions, it also adds startup notification support, for example. So if we wanted to use labwc as our compositor, we would need to get support for this in somehow, and it sounds as if that wouldn't be something you would want to upstream.

Have you considered a plugin interface or similar so that users can add additional functionality that you don't necessarily want to include as core labwc code?

**Comment by Consolatis** [comment]
> Have you considered a plugin interface or similar so that users can add additional functionality that you don't necessarily want to include as core labwc code?

Yes we considered that but we will not go that route in the foreseeable future. Our primary goal is to have a rock stable low-resource wayland compositor with reasonable Openbox and X11 feature support first and foremost. Once we deem labwc "stable"  enough we may think about plugin interfaces and similar.

**Comment by ghollingworth** [comment]
It would be useful to understand how this would best work if it was implemented and what capability it would be possible to achieve.  We really like your commitment to simplicity and "do one thing and do it well" and as such it really fits our software mentality.  We don't want to fork your repo if we can possibly avoid it and would be up for helping implement an interface that would allow us to support downstream functionality that you aren't interested in...  Even if it's an agreement about where we can patch your upstream releases to avoid conflicts...

Can email me to discuss gordon@raspberrypi.com

**Comment by emersion** [comment]
Note, various gtk-shell bits are upstream or being worked on. I would suggest spending time to try to get these standardized instead of spending time to implement the private gtk-shell protocol. For instance:

- Startup notifications are achieved via xdg-activation.
- Dialogs: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/228
- WM gestures: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/229
- Resize edge constraints: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/230
- Bell: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/231

**Comment by johanmalm** [comment]
We might be able to use `min_size` == `max_size` as a trigger to indicate that a window is "not resizable".

https://wayland.app/protocols/xdg-shell#xdg_toplevel:request:set_max_size

Do you have an application in mind that allows server-side-deco and also sets `min_size` to the same as `max_size`?

You can get the min/max size values with this patch:

```
diff --git a/src/xdg.c b/src/xdg.c
index 713eccd..000aaf5 100644
--- a/src/xdg.c
+++ b/src/xdg.c
@@ -127,6 +127,13 @@ handle_commit(struct wl_listener *listener, void *data)
 	if (update_required) {
 		view_impl_apply_geometry(view, size.width, size.height);
 	}
+
+	struct wlr_xdg_toplevel *toplevel = xdg_toplevel_from_view(view);
+	wlr_log(WLR_ERROR, "max_size=%dx%d; min_size=%dx%d",
+		toplevel->pending.max_width,
+		toplevel->pending.max_height,
+		toplevel->pending.min_width,
+		toplevel->pending.min_height);
 }
 
 static int
```

**Comment by stefonarch** [comment]
Maybe kcolorchooser is an example. On maximize it will jump to top left corner but not change size. Maybe disabling (removing) the button in this case?

**Comment by johanmalm** [comment]
> Maybe kcolorchooser is an example. On maximize it will jump to top left corner but not change size. Maybe disabling (removing) the button in this case?

Indeed a good choice! Have just tried. That gives us something to work with.
